design/yaml-versioning.md

```diff
@@ -0,0 +1,205 @@
+
+## Solution
+
+We introduce a new field, `version`, which lives at the root level of the pipeline.
+It's an integer whose valid values are (currently) only `0` and `1`.
+Any other number is a syntax error.
+If not present, the value is assumed to be `0` (unless included in a [template; see below](#templates)).
+
+We'll continue our commitment to backcompat within a single version.
+Whatever version is currently in active development will grow new features over time.
+Down-level versions will continue to be supported for the foreseeable future, though we reserve the right to remove versions eventually.
+
+### Version 0
+Version 0 is the current set of syntax, back-compat warts and all.
+No behavior changes are expected, and we'll freeze this version when v1 ships.
+No new features will land in v0.
+
+### Version 1
+Version 1 becomes our active development version.
+This spec doesn't attempt to list all the new things it will gain, since that's not predictable.
+Instead, it lists only the breaking changes we'll make now.
+
+#### Additions
+* `version: 1` must be specified at the root of the pipeline.
+
+#### Removals
+* Remove `phases`/`phase`. This became `jobs`/`job` with minimal other changes.
+* Remove `queue`. This includes all the nodes which moved under `strategy`.
+* Remove Git-LFS, shallow fetch, and other non-core properties from `repository`. These are now handled by `checkout`.
+* Remove sequence form of `resources`. This is completely handled by the current mapping form.
+* Remove `pr` and `schedule` nodes. (See [trigger changes](#trigger-changes) for their replacements.)
+* (Tentative) Remove `$[ ]` syntax, as we're planning to make `${{ }}` execute just-in-time.
+
+#### Trigger changes
+`trigger` becomes a parent of all `self`-repo triggers.
+CI keywords, which used to be directly under `trigger`, now live under a sub-entry called `push`.
+`pr` and `schedule`, which used to be pipeline-level, also live under `trigger`.
+Example:
+```yaml
+# old way
+trigger:
+  branches:
+  - master
+  - features/*
+pr:
+  branches:
+    include:
+    - master
+    - releases/*
+
+# new way
+trigger:
+  push:
+    branches:
+    - master
+    - features/*
+  pr:
+    branches:
+      include:
+      - master
+      - releases/*
+```
+
+`trigger: none` disables all triggers.
+Each trigger category (PR, CI, and Schedule) can be disabled independently:
+```yaml
+trigger:
+  pr: none
+  ci: ...
+```
+
+These changes also apply to `resource` triggers.
+
+#### Behavior changes
+* `repository` and `container` gain triggers by default. Users must suppress them using `trigger: none` if they don't want trigger behavior.
+* ...
+
+#### Syntax strictness
+* `repository` and `container` allow arbitrary properties today. In v1, we will fix the set of properties that are allowed and error if unknown properties are passed.
+* `task` inputs are arbitrary. In v1, only valid inputs (and aliases of inputs) will be accepted.
+* ...
+
+#### Still under discussion
```

Comment:
I'd love to, but I think it ends in absurdity. The simplest pipeline would be _required_ to read:
```yaml
stages:
- stage:
  jobs:
  - job:
    steps:
    - script: echo that's a lot of boilerplate
```

Instead, I'm trying to land a couple of inheritance behaviors that make it easier to grow up. A lot of the pain I've seen comes when you used to be able to put `pool` as a peer of `steps`, then tried to grow from one (implied) job to 2. If `pool` could live at any level (pipeline, stage, or job), then it would be legal to not indent the pool, and you'd get what you expect.